libraries/shared/storage/fetcher/geth_rpc_storage_fetcher.go

```diff
@@ -66,45 +61,177 @@ func (fetcher GethRpcStorageFetcher) FetchStorageDiffs(out chan<- types.RawDiff,
 		case err := <-clientSubscription.Err():
 			logrus.Errorf("error with client subscription: %s", err.Error())
 			errs <- err
-		case diff := <-ethStatediffPayloadChan:
-			logrus.Trace("received a statediff")
-			stateDiff := new(statediff.StateDiff)
-			decodeErr := rlp.DecodeBytes(diff.StateDiffRlp, stateDiff)
-			logrus.Tracef("received a statediff from block: %v", stateDiff.BlockNumber)
-			if decodeErr != nil {
-				logrus.Warn("Error decoding state diff into RLP: ", decodeErr)
-				errs <- decodeErr
+		case diffPayload := <-ethStatediffPayloadChan:
+			logrus.Trace("received a statediff payload")
+
+			switch fetcher.gethVersion {
+			case OldGethPatch:
+				fetcher.handleDiffsFromOldGethPatch(diffPayload, out, errs)
+			case NewGethPatchWithService:
+				fetcher.handleDiffsFromNewGethPatchWithService(diffPayload, out, errs)
+			case NewGethPatchWithFilter:
+				fetcher.handleDiffsFromNewGethPatchWithFilter(diffPayload, out, errs)
 			}
+		}
+	}
+}
 
-			accounts := getAccountsFromDiff(*stateDiff)
-			logrus.Trace(fmt.Sprintf("iterating through %d accounts on stateDiff for block %d", len(accounts), stateDiff.BlockNumber))
-			for _, account := range accounts {
-				logrus.Trace(fmt.Sprintf("iterating through %d Storage values on account", len(account.Storage)))
-				for _, accountStorage := range account.Storage {
-					diff, formatErr := fetcher.formatDiff(account, stateDiff, accountStorage)
-					logrus.Tracef("adding storage diff to out channel. keccak of address: %v, block height: %v, storage key: %v, storage value: %v",
-						diff.HashedAddress.Hex(), diff.BlockHeight, diff.StorageKey.Hex(), diff.StorageValue.Hex())
-					if formatErr != nil {
-						errs <- formatErr
-					}
-
-					out <- diff
-				}
+func (fetcher GethRpcStorageFetcher) handleDiffsFromOldGethPatch(payload filters.Payload, out chan<- types.RawDiff, errs chan<- error) {
+	stateDiff, decodeErr := fetcher.decodeStateDiffRlpFromPayloadForOldPatches(payload)
+	if decodeErr != nil {
+		errs <- decodeErr
```

Comment:
also not a blocker but do we potentially want to return after sending this error? Thinking we might not be able to do useful work afterword/might hit a null pointer error

Reply:
also this is even less of a blocker but maybe we could use `fmt.Errorf` to wrap the decode error and explain where he hit a snag (e.g. which patch)